### PR TITLE
Fix sphinx docs rendering of dispatched functions

### PR DESCRIPTION
**Pull request by eriknw**
@MridulS noticed the online docs no longer render properly for dispatched functions now that `_dispatch` is a class: https://github.com/networkx/networkx/pull/6840#issuecomment-1700779731

I'm not a Sphinx expert, but here is one way to fix the problem. It also allows the dispatch machinery to update the docstring. The signature of dispatched functions is unchanged in the online docs--that is, `backend=None, **backend_kwargs` is not shown.

### CODE REVIEW

**Comment by eriknw** [comment]
I don't know how to make `mypy` happy here. Or, rather, I don't think you would like my solution for making mypy happy 😜

**Comment by eriknw** [comment]
Btw, adding `__repr__` to `_dispatch` helps, but it doesn't _completely_ solve the issue. It results in e.g. `adjacency_matrix = ...` instead of `adjacency_matrix(...)`.

File: networkx/utils/backends.py [hunk 1]
```diff
@@ -817,3 +817,19 @@ def __reduce__(self):
...
+
+if os.environ.get("_NETWORKX_BUILDING_DOCS_"):
+    # When building docs with Sphinx, use the original function with the
+    # dispatched __doc__, b/c Sphinx renders normal Python functions better.
+    # This doesn't show e.g. `*, backend=None, **backend_kwargs` in the
+    # signatures, which is probably okay. It does allow the docstring to be
+    # updated based on the installed backends.
+    _orig_dispatch = _dispatch
+
+    def _dispatch(func=None, **kwargs):
```
Comment (MridulS) [inline networkx/utils/backends.py:830]:
```suggestion
    def _dispatch(func=None, **kwargs): # type: ignore
```
🙃

File: networkx/utils/backends.py [hunk 2]
```diff
@@ -817,3 +817,19 @@ def __reduce__(self):
...
+
+if os.environ.get("_NETWORKX_BUILDING_DOCS_"):
+    # When building docs with Sphinx, use the original function with the
+    # dispatched __doc__, b/c Sphinx renders normal Python functions better.
+    # This doesn't show e.g. `*, backend=None, **backend_kwargs` in the
+    # signatures, which is probably okay. It does allow the docstring to be
+    # updated based on the installed backends.
+    _orig_dispatch = _dispatch
+
+    def _dispatch(func=None, **kwargs): # type: ignore
```
Comment (MridulS) [inline networkx/utils/backends.py:830]:
```suggestion
    def _dispatch(func=None, **kwargs):  # type: ignore
```

**Pull request merged**